Add asset_allocations and outpoint_assets methods

zoedberg opened the pull request:
this PR is needed in order to complete the task

> Provide bindings for convinence functions [...]

defined in #2

N.B to merge this we need to do a release of rgb-node (containing LNP-BP/rgb-node#100)

zoedberg committed:
Add asset_allocations and outpoint_assets

dr-orlovsky commented on rust-lib/src/lib.rs:
```diff
@@ -346,3 +355,51 @@ pub extern "C" fn transfer(
+    let c_contract_id = unsafe { CStr::from_ptr(contract_id) };
+    let contract_id = ContractId::from_str(c_contract_id.to_str()?)?;
+
+    info!("AssetAllocationsArgs {{ contract_id: {:?} }}", contract_id);
```
Don't know why Alekos used `info!` in these functions here, this is clearly a `debug!` type of information

dr-orlovsky replied:
Seems like this is still have to be `debug!`, as was discussed regarding the criteria. This is read-only method, so no need to provide verbose output about it's args

zoedberg replied:
as already agreed, I will change this when facing #23

dr-orlovsky commented on rust-lib/src/lib.rs:
```diff
@@ -346,3 +355,51 @@ pub extern "C" fn transfer(
+    let c_outpoint = unsafe { CStr::from_ptr(outpoint) };
+    let outpoint = OutPoint::from_str(c_outpoint.to_str()?)?;
+
+    info!("OutpointAssets {{ outpoint: {:?} }}", outpoint);
```
Ibid

zoedberg replied:
as above

dr-orlovsky commented on rust-lib/src/lib.rs:
```diff
@@ -346,3 +355,51 @@ pub extern "C" fn transfer(
+    let runtime = Runtime::from_opaque(runtime)?;
+
+    let c_contract_id = unsafe { CStr::from_ptr(contract_id) };
+    let contract_id = ContractId::from_str(c_contract_id.to_str()?)?;
```
While `from_str` probably uses Bech32 encoding internally, previously there was an alternative form of `ContractId` representation as a hex string. So I propose to do more explicit conversion using `from_bech32_str` method (which also provides more clear error reporting) https://github.com/LNP-BP/rust-lnpbp/blob/master/src/rgb/bech32.rs#L219

zoedberg replied:
:+1: done

dr-orlovsky reviewed (changes requested):
Can we just switch to using github master branch in Cargo for now?

zoedberg commented:
> Can we just switch to using github master branch in Cargo for now?

We can, but I'd really prefer to not do that. If something breaks in rgb-node `master` (as it's often the case), it would cause this project to break too. If you don't want to do a new release on crates.io we can add a tag on rgb-node and use that (my point is having the sdk always sticking to a fixed version in time).

zoedberg committed:
Use from_bech32_str to parse contract_id

zoedberg committed:
rust-lib: use log macro

dr-orlovsky commented:
Version 0.2-beta.3 is published: https://crates.io/crates/rgb_node/0.2.0-beta.3 so is this PR is ready to be merged? Can you pls update it to use the new published version so CI will success?

zoedberg commented:
> Version 0.2-beta.3 is published: https://crates.io/crates/rgb_node/0.2.0-beta.3 so is this PR is ready to be merged? Can you pls update it to use the new published version so CI will success?

I've changed rgb-node version and rebased on master.
Also added fix to bindings workflow (as done for rgb-node yesterday).

But I'm not sure this is ready to be merged. Shouldn't the functions return the response data to the user? Should I use JSON as a serialization format of the response? For now there are no other functions in the sdk that output a response. @dr-orlovsky

dr-orlovsky commented on rust-lib/src/lib.rs:
```diff
@@ -346,3 +355,51 @@ pub extern "C" fn transfer(
+
+    info!("AssetAllocationsArgs {{ contract_id: {:?} }}", contract_id);
+
+    runtime.asset_allocations(contract_id)?;
+
+    Ok(())
```
just
```suggestion
    runtime.asset_allocations(contract_id)
```

dr-orlovsky commented on rust-lib/src/lib.rs:
```diff
@@ -346,3 +355,51 @@ pub extern "C" fn transfer(
+
+    info!("OutpointAssets {{ outpoint: {:?} }}", outpoint);
+
+    runtime.outpoint_assets(outpoint)?;
+
+    Ok(())
```
Just
```suggestion
    runtime.outpoint_assets(outpoint)
```

dr-orlovsky reviewed (changes requested):
You are absolutely right, overlooked again that we need to return data. These are read-only methods doing nothing then providing the client with the information.

For reading data we already have a "main" approach: to access the cache directly, which is stored either as a single YAML/JSON file or SQLite database and read data from there. So the whole reason of having this functions exported via FFI is to return specially-constructed  data structure native for each specific language. Since we have to pass it through C, I propose to do convert a return type into an array of some cuctom-defined C struct, and then process it on the language side to convert into a lang-specific struct (and it will be JSON object in JavaScript, but something else in Java, Swift etc).

```c
struct allocation_list {
   uint len;
   struct allocation *allocations;
}

struct allocation {
   // ...
}
```

Please let me know if this seems to complicated for you and then I will pick up this thing or ask @rajarshimaitra to help.

zoedberg commented:
> Since we have to pass it through C, I propose to do convert a return type into an array of some cuctom-defined C struct, and then process it on the language side to convert into a lang-specific struct (and it will be JSON object in JavaScript, but something else in Java, Swift etc).

Sincerely, I don't think this is the best approach. I don't see any reasons for doing conversion in C. It would mean doing a lot of work on each language binding and giving to each C interface the knowledge of the rust structure, it seems useless duplication to me.

Moreover, JSON is the most common serialization format (especially for APIs), it can be used by almost all existing languages (for sure those we want to support in the sdk) and returning the JSON serialized string directly from rust would mean zero work on the binding interfaces (in rust we already know how to serialize objects to strings).

If you don't agree with this, I'd like to know the advantages of your approach

dr-orlovsky commented:
I agree with your JSON analysis, but for it we do not need these functions, it's already there :)

zoedberg commented:
> I agree with your JSON analysis, but for it we do not need these functions, it's already there :) Just read the asset file from data/cache directory directly and that's all, no other integration is needed.

For me the point of having an sdk is to provide rgb APIs to native languages, by lifting up the user from the hassle of creating a zmq connection (supposing an official libzmq support for their language exists) and parsing the input/output data into the format requested by the rgb-node. I think that for consistency, whatever the user wants to do and whatever the APIs do, they all should be used in the same manner and the response be always in the same format.

So, even if for data read operations we could just tell the user to access the cache (reading a file or a DB), I think that this would result in a inconsistency with other APIs, an hassle for the sdk user, and also an hurdle in case the user wants to use the node remotely.

Therefore, for me, in the sdk it makes sense to always call the APIs from the `i9n` module, and in case the API has data to return to the user, it should always use the same format, i.e. JSON.

dr-orlovsky commented:
Ok.

Personally for me "overhead" of creating another struct is so tiny (few mins) that it is not even worth discussing, since it's faster just to do it. Thus I always prefer native things, even if they require some more code — and JSON looks like extremely unefficient, unsecure and type unsafe thing which I am trying to avoid at all costs. However I appreciate your point of view in terms of those who will use this API (instead of working at the low level or writing some rust + ZMQ code), so whatever

zoedberg commented:
I see a third possible approach.

But first, a recap of the proposed solutions, as I understood them:
- your approach requires to manually convert Rust structures to C, then to C++ (handled by swig) and finally to each language supported by this sdk
- my approach requires to convert Rust structures to a JSON string, that will be forwarded (as is) from C through C++ and to the native language. This option also allows conversion from JSON to native structures on the native language side.

Third possible approach: pass data in binary form (just like zmq does) and then parse it back to the native language (using native language code, not C).
Bonus: this would result in code that can be also used by developers that prefer to handle the connection to zmq directly, since they would still need to know how to parse the binary blob into a native structure.

I tried to be brief, we can discuss it further if needed.

dr-orlovsky commented:
Yes, the binary form is a good option, but then we need to write binary decoders for each data structure in each of the languages, which for me seems larger effort than in-memory struct passing as args.

As I said, I'm fine with doing JSON for now since I do not have time to work on this issue myself.

zoedberg commented:
> Yes, the binary form is a good option, but then we need to write binary decoders for each data structure in each of the languages, which for me seems larger effort than in-memory struct passing as args.

I'm pretty convinced that binary decoders do not require more code or effort than in-memory struct passing.

It's true that the binary solution requires to write a decode method for each struct, but on the rust side the conversion of the object to a binary blob doesn't require extra code.

Instead, for the in-memory struct solution we need to define C structs in the rust-lib (one for each data structure), define their conversions in each swig interface (usually one for language) and then helpers method on each native language the sdk supports.

Moreover, since the binary solution also covers a second use case (the one of users handling the zmq connection theirselves), the time required for it should count half (since we would kill two birds with one stone) ;)

> As I said, I'm fine with doing JSON for now since I do not have time to work on this issue myself.

I agree that to close this PR quickly we can just use JSON, since it doesn't require any extra code. I can do that and then open a dedicated issue to not forget to change this temporary solution to a more definitive one

UkolovaOlga commented:
Taking into account the discussion above, @zoedberg feel free to proceed with JSON.

zoedberg committed:
rust-lib: return outuput as json

zoedberg committed:
ffi nodejs: add CResultString conversion to swig interface

zoedberg committed:
ffi android: add CResultString conversion to swig interface

zoedberg committed:
ffi android: add helper methods to Runtime class

zoedberg committed:
demo nodejs: add outpoint_assets call + fixes/improvements

zoedberg committed:
demo android: add outpoint_assets call

dr-orlovsky commented on demo/nodejs/example.js:
```diff
@@ -21,17 +31,34 @@ const transferData = {
+    })
+    /*
+    .then(() => {
+        return ex.transfer(runtime, JSON.stringify(transferData.inputs), JSON.stringify(transferData.allocate),
```
Why we have commented out the transfer function from the demo?

dr-orlovsky replied:
Anyway I think the correct place to add it back will be #21 so merging this version here

zoedberg replied:
@dr-orlovsky as reported on #14, currently `transfer` is not working, so for now it's better to leave it commented. I think we should add both `import` and `invoice` methods (both on `rgb::i9n` and in the sdk). if you agree on this I'll open the related issues

dr-orlovsky replied:
Agree

dr-orlovsky reviewed (approved):
Looks good to be merged! The only question is why we removed transfer from the demo?
